includes/data-stores/class-wc-customer-download-log-data-store.php

```diff
@@ -0,0 +1,196 @@
+<?php
+if ( ! defined( 'ABSPATH' ) ) {
+	exit;
+}
+
+/**
+ * WC Customer Download Log Data Store.
+ *
+ * @version  3.3.0
+ * @category Class
+ * @author   WooThemes
+ */
+class WC_Customer_Download_Log_Data_Store implements WC_Customer_Download_Log_Data_Store_Interface {
+
+	/**
+	 * Create download log entry.
+	 *
+	 * @param WC_Customer_Download_Log $download_log
+	 */
+	public function create( &$download_log ) {
+		global $wpdb;
+
+		// Always set a timestamp
+		if ( is_null( $download_log->get_timestamp( 'edit' ) ) ) {
+			$download_log->set_timestamp( current_time( 'timestamp', true ) );
+		}
+
+		$data = array(
+			'timestamp'           => date( 'Y-m-d H:i:s', $download_log->get_timestamp( 'edit' )->getTimestamp() ),
+			'permission_id'       => $download_log->get_permission_id( 'edit' ),
+			'user_id'             => $download_log->get_user_id( 'edit' ),
+			'user_ip_address'     => $download_log->get_user_ip_address( 'edit' ),
+		);
+
+		$format = array(
+			'%s',
+			'%s',
+			'%s',
+			'%s',
+		);
+
+		$result = $wpdb->insert(
+			$wpdb->prefix . 'woocommerce_downloadable_product_download_log',
+			apply_filters( 'woocommerce_downloadable_product_download_log_data', $data ),
+			apply_filters( 'woocommerce_downloadable_product_download_log_format', $format, $data )
+		);
+
+		do_action( 'woocommerce_downloadable_product_download_log', $data );
+
+		if ( $result ) {
+			$download_log->set_id( $wpdb->insert_id );
+			$download_log->apply_changes();
+		}
+	}
+
+	/**
+	 * Method to read a download log from the database.
+	 *
+	 * @param $download_log
+	 *
+	 * @throws Exception
+	 */
+	public function read( &$download_log ) {
+		global $wpdb;
+
+		$download_log->set_defaults();
+
+		if ( ! $download_log->get_id() || ! ( $raw_download_log = $wpdb->get_row( $wpdb->prepare( "SELECT * FROM {$wpdb->prefix}woocommerce_downloadable_product_download_log WHERE download_log_id = %d", $download_log->get_id() ) ) ) ) {
```

Comment:
This is a really long `if`; would be good to break it up. (Generally, it's preferred not to do too much computation within `if` clauses anyway; better to assign to a var and check the var.)